src/lib.rs

```diff
@@ -0,0 +1,459 @@
+            let contract_parameter = ContractParameter::<T::Hash, T::AccountId> {
+                uid: uid,
+                token_id: token_id,
+                token_owner: sender.clone(),
+                metadata: metadata.clone(),
+                proof_leaves: proof_leaves,
+            };
+
+            // Append the parameter after method
+            call.append(&mut Encode::encode(&contract_parameter));
+
+            // Call the contract via bare call
+            <contracts::Module<T>>::bare_call(
+                sender,
+                Self::validator_of(uid).unwrap(),
+                value,
+                gas_limit,
+                call);
+
+            Ok(())
+        }
+
+        // Call back interface for smart contract
+        fn finish_mint(origin, uid: RegistryUid, token_id: T::Hash, token_owner: T::AccountId, metadata: Vec<u8>) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+
+            // Ensure uid is existed
+            Self::ensure_sender_is_validation_function(uid, &sender)?;
+
+            // Use the uid to create a new ERC721 token, BUG owner is not sender.
+            Self::mint_nft(&token_owner, &token_id)?;
+
+            // Get storage fee for metadata
+            let total_deposit = Self::compute_metadata_fee(metadata.len() as u32);
+
+            // Reserve fee for token
+            <T as Trait>::Currency::reserve(&token_owner, total_deposit)?;
+
+            // Insert token id to registry id map
+            <TokenByRegistryId<T>>::insert(&token_id, uid);
+
+            // Insert token metadata
+            <RegistryTokenMetadata<T>>::insert(&token_id, metadata);
+
+            // Just emit an event
+            Self::deposit_event(RawEvent::MintNft(uid, token_id));
+
+            Ok(())
+        }
+
+        // transfer_from will transfer to addresses even without a balance
+        fn transfer_from(origin, from: T::AccountId, to: T::AccountId, token_id: T::Hash) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+
+            // Check if sender can transfer token
+            Self::ensure_is_approved_or_owner(&token_id, &sender)?;
+
+            // Call transfer
+            Self::_transfer_from(&from, &to, &token_id)?;
+
+            // Compute the total deposit fee
+            let total_deposit = Self::get_metadata_fee(&token_id);
+
+            // Reserve the currency
+            <T as Trait>::Currency::repatriate_reserved(&from, &to, total_deposit)?;
+
+            Ok(())
+        }
+
+        // Burn the token and unreserve the currency for metadata
+        fn burn(origin, token_id: T::Hash) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+
+            // Ensure token is existed
+            Self::ensure_token_exists(&token_id)?;
+
+            // Check if sender can transfer token
+            Self::ensure_is_approved_or_owner(&token_id, &sender)?;
+
+            // burn the token
+            Self::_burn(&token_id)?;
```

Comment:
1. we can remove metadata and registry id for the burned token.
2. I am wondering if your question is if user burn token in ERC721 module. If that is the question, I think we don't expose the burn interface in ERC721. I also removed the transfer interface in ERC721 because we need put the currency reserver/unreserve in mint/transfer/burn.

Reply:
2. ERC721 implementation should be generic (so it can be reused by others), meaning the that we need to expose all the functions, including burn, transfer.
1. Accordingly, the NFT registry needs a way to transfer and remove state even if the user did transfer/burn with the ERC721 module.

Let me clarify this before you change anything here.

Reply:
I also realize the dilemma. If we expose transfer/burn in ERC721 pallet, and reuse the pallet in NFT, we can't block user call them via ERC721 interface. Other way is put all these storage and methods into single pallet, but seems nft and ERC 721 are coupled.

Reply:
The Substrate way to do this is to specify a handler type on the Trait that can be triggered, you can have a look t how elections and collective pallets interact.